docs/proposals/dynamic-app-parameters.md

```diff
@@ -0,0 +1,181 @@
+
+In a typical Argo CD installation however, the component that’s used to render Helm charts (the repository server) does not have access to nor contextual knowledge (beside a few informational assets such cluster version and APIs provided) about the cluster where the chart is going to be installed to. Argo CD performs an offline templating of the chart, so there is no way for this kind of dynamic parameterization to work.
+
+The current workaround is to parameterize all Argo CD Applications using explicit, application specific parameters. This can become tedious and cumbersome, especially when using mechanisms such as ApplicationSet to render large amounts of Applications. Also, secret data is not to be stored in Application manifests, since these objects are considered non-confidential.
+
+The community wants to leverage the functionality behind `helm lookup` with Argo CD. However natively supporting `helm lookup` would require a significant architectural change, and it would also have severe security implications to be resolved.
+
+## Non-goals
+
+* Supporting `helm lookup` “as is” is a non-goal. Example, `helm lookup` constructs like `“lookup "v1" "Pod" "mynamespace" "mypod"` are not supported. This proposal addresses an Argo CD-native mechanism to support consumption of in-cluster information in helm charts and possibly other types of Applications in the future. 
+* Emulate or otherwise make this feature a replacement for secrets management (albeit it may allow Applications to read values stored in secrets on the cluster)
+
+## Proposal
+
+The authors of this document propose the introduction of a new kind of parameter for Helm applications (and later possibly other types of Applications): The dynamic parameter.
+
+Currently, Helm Applications within Argo CD support _statically defined_ parameters, which are simple key/value mappings and which need to be set to the desired value before the Application's manifests are being rendered.
+
+A dynamic parameter is defined similarly to how a static parameter is defined currently, but instead of supplying a static value, it is configured using a reference to a resource that exists in the destination cluster. 
+
+For example:
+
+```yaml
+kind: Application
+apiVersion: argoproj.io/v1alpha1
+metadata:
+  name: foo-app
+spec:
+  project: default
+  source:
+    repoURL: https://some.helm.repo.com
+    chart: foo
+    targetRevision: 1.0.x
+    helm:
+      parameters:
+      - name: foo
+        value: bar
+      dynamicParameters:
+      - name: bar
+        forceString: true # optional, will use --set-string
+        resourceRef:
+          kind: ConfigMap
+          group: ""
```

Comment:
Would be great if `group` was optional

Reply:
Do you mean like, if `group` is not specified, it automatically will assume the core API group?

Reply:
IMHO that's be good as it would reduce the quantity of "boilerplate" code.
Another variant would be to automatically take the matching group if there is no risk of duplicates. I.e. If resource XYZ is only present in group ABC, then it would work. In case of duplicates a warning/error would arise.

Reply:
+1 on "if `group` is omitted, assume the core API".